spring-cloud-gateway-server-mvc/src/main/java/org/springframework/cloud/gateway/server/mvc/filter/BeforeFilterFunctions.java

```diff
@@ -350,9 +353,11 @@ public static Function<ServerRequest, ServerRequest> setPath(String path) {
 		return request -> {
 			Map<String, Object> uriVariables = MvcUtils.getUriTemplateVariables(request);
 			URI uri = uriTemplate.expand(uriVariables);
-			String newPath = uri.getRawPath();
 
-			URI prefixedUri = UriComponentsBuilder.fromUri(request.uri()).replacePath(newPath).build().toUri();
+			URI prefixedUri = UriComponentsBuilder.fromUri(request.uri())
+				.replacePath(uri.getRawPath())
+				.build(true)
+				.toUri();
```

Comment:
The modified path is already encoded, so no additional encoding will be applied. 
Furthermore, other parts, such as request parameters, are not affected.